### PR TITLE
feat: add ignoreExportsUsedInFile option

### LINKED ISSUE #127: Allow ignoring unused exported types that are used by something exported
In TypeScript-land, I like exporting types corresponding to used values, even if those types aren't directly used by anything.

For example, `greet` here would be allowed but `GreetOptions` would be complained about:

```ts
// greet.ts
export interface GreetOptions {
  message: string;
  repeat?: number;
}

export function greet(options: GreetOptions) {
  console.log(options);
}
```

```ts
// index.ts
import { greet } from "./greet";

greet({ message: "Hello, world!" });
```

```plaintext
$ knip

Unused exported types (1)
GreetOptions  interface  greet.ts
```

Could we have an option to allow these used-indirectly-but-not-directly types please? I'd be happy to send a PR!

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #127.

### CODE REVIEW

File: src/ConfigurationChief.ts [hunk 1]
```diff
@@ -51,6 +51,7 @@ const defaultConfig: Configuration = {
   ignore: [],
   ignoreBinaries: [],
   ignoreDependencies: [],
+  ignoreExportsUsedInFile: false,
```
Comment (JoshuaKGoldberg) [inline src/ConfigurationChief.ts:54]: This feels awfully specific for a high-level option, but I couldn't think of a better way to position it... 😕

**Comment by webpro** [comment]
Thanks, Josh! Looks good at first sight. Will play a bit with it tomorrow. Saw a dangling `console.log` :)

**Comment by JoshuaKGoldberg** [comment]
Ah, dangit :smile: thanks, removed.

File: tests/exports-in-file.test.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import baseArguments from './helpers/baseArguments.js';
+import baseCounters from './helpers/baseCounters.js';
+
+const cwd = resolve('tests/fixtures/exports-in-file');
+
+test('Find unused exports respecting ignoreExportsUsedInFile', async () => {
+  const { issues, counters } = await main({
+    ...baseArguments,
+    cwd,
+  });
```
Comment (webpro) [inline tests/exports-in-file.test.ts:14]: Shouldn't this include the `ignoreExportsUsedInFile: true` option?

**Review by webpro — COMMENTED** [review]
From the initial RFC I was under the impression this flag would only be effective for types. Including exports/values is quite a difference, think tree-shaking (types are always removed).

I see the value of both, but I think it should be possible to allow this feature only for types. Which doesn't help the option/naming challenge 😅

What do you think?

File: tests/exports-in-file.test.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [reply tests/exports-in-file.test.ts]: Ah, the tests weren't testing the way I thought they were. Sorry about that. Should be fixed now.

File: src/ConfigurationValidator.ts [hunk 3]
```diff
@@ -34,6 +34,22 @@ const rootConfigurationSchema = z.object({
...
+          z.literal('enum'),
+          z.literal('function'),
+          z.literal('interface'),
+          z.literal('member'),
+          z.literal('type'),
+        ]),
+        z.boolean()
+      ),
+    ])
+    .optional(),
```
Comment (JoshuaKGoldberg) [inline src/ConfigurationValidator.ts:52]: > possible to allow this feature only for types

Makes sense! I based these values on `SymbolType` (`exportedItem.type`), excluding `'unknown'`. WDYT?

**Comment by webpro** [comment]
This is great! I love this config granularity. Thanks Joshua, much appreciated! I'm seeing some really minor things I might want to refactor later, but it's just implementation details. Happy to merge.

**Comment by webpro** [comment]
:rocket: _This pull request is included in v2.13.0. See [Release 2.13.0](https://github.com/webpro/knip/releases/tag/2.13.0) for release notes._

**Comment by JoshuaKGoldberg** [comment]
Yay, thanks! I'm excited to get a first big new change here 😄

**Pull request merged**